File: packages/shared/protocol.py
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field


class TranscriptSegment(BaseModel):
    id: str
    timestamp: datetime
    text: str
    speaker: Optional[str] = None
    confidence: Optional[float] = None


class ContextCard(BaseModel):
    id: str
    type: Literal["knowledge", "summary", "action"]
    title: str
    body: str
    sources: list[str] = Field(default_factory=list)
    relevance_score: float = 0.0


class SessionSummary(BaseModel):
    id: str
    started_at: datetime
    transcript_segments: list[TranscriptSegment] = Field(default_factory=list)
    context_cards: list[ContextCard] = Field(default_factory=list)


class AudioChunkMessage(BaseModel):
    type: Literal["audio_chunk"] = "audio_chunk"
    data: str
    seq: int
    mime_type: Optional[str] = "audio/webm"


class ClientMessage(BaseModel):
    type: str

File: packages/shared/protocol.ts
export type ContextCardType = "knowledge" | "summary" | "action";

export interface TranscriptSegment {
  id: string;
  timestamp: string;
  text: string;
  speaker?: string | null;
  confidence?: number | null;
}

export interface ContextCard {
  id: string;
  type: ContextCardType;
  title: string;
  body: string;
  sources: string[];
  relevance_score: number;
}

export interface SessionSummary {
  id: string;
  started_at: string;
  transcript_segments: TranscriptSegment[];
  context_cards: ContextCard[];
}

export type ClientMessage =
  | { type: "audio_chunk"; data: string; seq: number; mime_type?: string }
  | { type: "ping" };

export type ServerMessage =
  | { type: "transcript"; segment: TranscriptSegment }
  | { type: "context"; cards: ContextCard[]; summary?: string; decisions?: string[]; open_questions?: string[] }
  | { type: "status"; message: string }
  | { type: "error"; message: string }
  | { type: "pong" };
